# Урок 2. Обработка данных для анализа

In [1]:
library(cluster) # install.packages("cluster")

In [2]:
?votes.repub

votes.repub              package:cluster               R Documentation

_V_o_t_e_s _f_o_r _R_e_p_u_b_l_i_c_a_n _C_a_n_d_i_d_a_t_e _i_n _P_r_e_s_i_d_e_n_t_i_a_l _E_l_e_c_t_i_o_n_s

_D_e_s_c_r_i_p_t_i_o_n:

     A data frame with the percents of votes given to the republican
     candidate in presidential elections from 1856 to 1976.  Rows
     represent the 50 states, and columns the 31 elections.

_U_s_a_g_e:

     data(votes.repub)
     
_S_o_u_r_c_e:

     S. Peterson (1973): _A Statistical History of the American
     Presidential Elections_.  New York: Frederick Ungar Publishing Co.

     Data from 1964 to 1976 is from R. M. Scammon, _American Votes 12_,
     Congressional Quarterly.

In [2]:
head(votes.repub)

X1856 X1860 X1864 X1868 X1872 X1876 X1880 X1884 X1888 X1892 ...
Alabama       NA    NA    NA 51.44 53.19 40.02 36.98 38.44 32.28  3.95 ...
Alaska        NA    NA    NA    NA    NA    NA    NA    NA    NA    NA ...
Arizona       NA    NA    NA    NA    NA    NA    NA    NA    NA    NA ...
Arkansas      NA    NA    NA 53.73 52.17 39.88 39.55 40.50 38.07 32.01 ...
California 18.77 32.96 58.63 50.24 56.38 50.88 48.92 52.08 49.95 43.76 ...
Colorado      NA    NA    NA    NA    NA    NA 51.28 54.39 55.31 41.13 ...
           X1940 X1944 X1948 X1952 X1956 X1960 X1964 X1968 X1972 X1976
Alabama    14.34 18.20 19.04 35.02 39.39 41.75 69.5  14.0  72.4  43.48
Alaska        NA    NA    NA    NA    NA 50.94 34.1  45.3  58.1  62.91
Arizona    36.01 40.90 43.82 58.35 60.99 55.52 50.4  54.8  64.7  58.62
Arkansas   20.87 29.84 21.02 43.76 45.82 43.06 43.9  30.8  68.9  34.97
California 41.35 42.99 47.14 56.39 55.40 50.10 40.9  47.8  55.0  50.89
Colorado   50.92 53.21 46.52 60.27 59.49 54.63 38.7  50.5  62.6  55.89

In [4]:
str(votes.repub)

'data.frame':	50 obs. of  31 variables:
 $ X1856: num  NA NA NA NA 18.8 ...
 $ X1860: num  NA NA NA NA 33 ...
 $ X1864: num  NA NA NA NA 58.6 ...
 $ X1868: num  51.4 NA NA 53.7 50.2 ...
 $ X1872: num  53.2 NA NA 52.2 56.4 ...
 $ X1876: num  40 NA NA 39.9 50.9 ...
 $ X1880: num  37 NA NA 39.5 48.9 ...
 $ X1884: num  38.4 NA NA 40.5 52.1 ...
 $ X1888: num  32.3 NA NA 38.1 50 ...
 $ X1892: num  3.95 NA NA 32.01 43.76 ...
 $ X1896: num  28.1 NA NA 25.1 49.1 ...
 $ X1900: num  34.7 NA NA 35 54.5 ...
 $ X1904: num  20.6 NA NA 40.2 61.9 ...
 $ X1908: num  24.4 NA NA 37.3 55.5 ...
 $ X1912: num  8.26 NA 12.74 19.73 0.58 ...
 $ X1916: num  22 NA 35.4 28 46.3 ...
 $ X1920: num  31 NA 55.4 38.7 66.2 ...
 $ X1924: num  27 NA 41.3 29.3 57.2 ...
 $ X1928: num  48.5 NA 57.6 39.3 64.7 ...
 $ X1932: num  14.2 NA 30.5 12.9 37.4 ...
 $ X1936: num  12.8 NA 26.9 17.9 31.7 ...
 $ X1940: num  14.3 NA 36 20.9 41.4 ...
 $ X1944: num  18.2 NA 40.9 29.8 43 ...
 $ X1948: num  19 NA 43.8 21 47.1 ...
 $ X1952: num 

In [5]:
votes.repub[1:6, 1:10]

X1856 X1860 X1864 X1868 X1872 X1876 X1880 X1884 X1888 X1892
Alabama       NA    NA    NA 51.44 53.19 40.02 36.98 38.44 32.28  3.95
Alaska        NA    NA    NA    NA    NA    NA    NA    NA    NA    NA
Arizona       NA    NA    NA    NA    NA    NA    NA    NA    NA    NA
Arkansas      NA    NA    NA 53.73 52.17 39.88 39.55 40.50 38.07 32.01
California 18.77 32.96 58.63 50.24 56.38 50.88 48.92 52.08 49.95 43.76
Colorado      NA    NA    NA    NA    NA    NA 51.28 54.39 55.31 41.13

In [8]:
head(colMeans(votes.repub))   # rowMeans()

X1856 X1860 X1864 X1868 X1872 X1876 
   NA    NA    NA    NA    NA    NA

In [9]:
class(colMeans(votes.repub))

[1] "numeric"

In [7]:
head(colMeans(votes.repub, na.rm = T))

X1856    X1860    X1864    X1868    X1872    X1876 
39.46850 44.58826 57.88080 54.13697 57.14622 48.49378

In [23]:
dim(votes.repub)[1]

[1] 50

### Операторы if else, ifelse

In [24]:
if(mean(votes.repub[, 30]) > 60){   # в 30-м столбце нет пропущенных значений
    print("республиканцы набрали высокий процент голосов")
}else{
    print("республиканцы набрали менее 60% голосов")
}

[1] "республиканцы набрали высокий процент голосов"


In [25]:
if(mean(votes.repub[, 7], na.rm = T) > 60){   # в 7-м столбце есть пропущенные значения
    print("республиканцы набрали высокий процент голосов")
}else{
    print("республиканцы набрали менее 60% голосов")
}

[1] "республиканцы набрали менее 60% голосов"


In [26]:
?ifelse

ifelse                  package:base                   R Documentation

_C_o_n_d_i_t_i_o_n_a_l _E_l_e_m_e_n_t _S_e_l_e_c_t_i_o_n

_D_e_s_c_r_i_p_t_i_o_n:

     'ifelse' returns a value with the same shape as 'test' which is
     filled with elements selected from either 'yes' or 'no' depending
     on whether the element of 'test' is 'TRUE' or 'FALSE'.

_U_s_a_g_e:

     ifelse(test, yes, no)
     
_A_r_g_u_m_e_n_t_s:

    test: an object which can be coerced to logical mode.

     yes: return values for true elements of 'test'.

      no: return values for false elements of 'test'.

_D_e_t_a_i_l_s:

     If 'yes' or 'no' are too short, their elements are recycled.
     'yes' will be evaluated if and only if any element of 'test' is
     true, and analogously for 'no'.

     Missing values in 'test' give missing values in the result.

_V_a_l_u_e:

     A vector of the same length and attributes (including dimensions
     and '"class"') as 'test' and data values from the values of 'yes'
     or 'no'.  The mode of the answer will be coerced from logical to
     accommodate first any values taken from 'yes' and then any values
     taken from 'no'.

_W_a_r_n_i_n_g:

     The mode of the result may depend on the value of 'test' (see the
     examples), and the class attribute (see 'oldClass') of the result
     is taken from 'test' and may be inappropriate for the values
     selected from 'yes' and 'no'.

     Sometimes it is better to use a construction such as
     
       (tmp <- yes; tmp[!test] <- no[!test]; tmp)
     , possibly extended to handle missing values in 'test'.

     Further note that 'if(test) yes else no' is much more efficient
     and often much preferable to 'ifelse(test, yes, no)' whenever
     'test' is a simple true/false result, i.e., when 'length(test) ==
     1'.

     The 'srcref' attribute of functions is handled specially: if
     'test' is a simple true result and 'yes' evaluates to a function
     with 'srcref' attribute, 'ifelse' returns 'yes' including its
     attribute (the same applies to a false 'test' and 'no' argument).
     This functionality is only for backwards compatibility, the form
     'if(test) yes else no' should be used whenever 'yes' and 'no' are
     functions.

_R_e_f_e_r_e_n_c_e_s:

     Becker, R. A., Chambers, J. M. and Wilks, A. R. (1988) _The New S
     Language_.  Wadsworth & Brooks/Cole.

_S_e_e _A_l_s_o:

     'if'.

_E_x_a_m_p_l_e_s:

     x <- c(6:-4)
     sqrt(x)  #- gives warning
     sqrt(ifelse(x >= 0, x, NA))  # no warning
     
     ## Note: the following also gives the warning !
     ifelse(x >= 0, sqrt(x), NA)
     
     
     ## ifelse() strips attributes
     ## This is important when working with Dates and factors
     x <- seq(as.Date("2000-02-29"), as.Date("2004-10-04"), by = "1 month")
     ## has many "yyyy-mm-29", but a few "yyyy-03-01" in the non-leap years
     y <- ifelse(as.POSIXlt(x)$mday == 29, x, NA)
     head(y) # not what you expected ... ==> need restore the class attribute:
     class(y) <- class(x)
     y
     ## This is a (not atypical) case where it is better *not* to use ifelse(),
     ## but rather the more efficient and still clear:
     y2 <- x
     y2[as.POSIXlt(x)$mday != 29] <- NA
     ## which gives the same as ifelse()+class() hack:
     stopifnot(identical(y2, y))
     
     
     ## example of different return modes (and 'test' alone determining length):
     yes <- 1:3
     no  <- pi^(1:4)
     utils::str( ifelse(NA,    yes, no) ) # logical, length 1
     utils::str( ifelse(TRUE,  yes, no) ) # integer, length 1
     utils::str( ifelse(FALSE, yes, no) ) # double,  length 1

In [27]:
x<-c(3, 0, 0, 0, 1, 0)
x

[1] 3 0 0 0 1 0

In [28]:
ifelse(x!=0, "Yes", "No")

[1] "Yes" "No"  "No"  "No"  "Yes" "No"

В том случае ,если среднее значение доли голосов, отданных за республиканцев, по штатам более 60, будет выводиться сообщение : "республиканцы набрали менее 60% голосов", в противном случае "республиканцы набрали менее 60% голосов"

In [29]:
ifelse(colMeans(votes.repub,na.rm = TRUE)>60, 
       "республиканцы набрали высокий поцент голосов","республиканцы набрали менее 60% голосов")

X1856 
     "республиканцы набрали менее 60% голосов" 
                                         X1860 
     "республиканцы набрали менее 60% голосов" 
                                         X1864 
     "республиканцы набрали менее 60% голосов" 
                                         X1868 
     "республиканцы набрали менее 60% голосов" 
                                         X1872 
     "республиканцы набрали менее 60% голосов" 
                                         X1876 
     "республиканцы набрали менее 60% голосов" 
                                         X1880 
     "республиканцы набрали менее 60% голосов" 
                                         X1884 
     "республиканцы набрали менее 60% голосов" 
                                         X1888 
     "республиканцы набрали менее 60% голосов" 
                                         X1892 
     "республиканцы набрали менее 60% голосов" 
                                         X1896 
     "республиканцы набрали менее 60% голосов" 
                                         X1900 
     "республиканцы набрали менее 60% голосов" 
                                         X1904 
     "республиканцы набрали менее 60% голосов" 
                                         X1908 
     "республиканцы набрали менее 60% голосов" 
                                         X1912 
     "республиканцы набрали менее 60% голосов" 
                                         X1916 
     "республиканцы набрали менее 60% голосов" 
                                         X1920 
     "республиканцы набрали менее 60% голосов" 
                                         X1924 
     "республиканцы набрали менее 60% голосов" 
                                         X1928 
     "республиканцы набрали менее 60% голосов" 
                                         X1932 
     "республиканцы набрали менее 60% голосов" 
                                         X1936 
     "республиканцы набрали менее 60% голосов" 
                                         X1940 
     "республиканцы набрали менее 60% голосов" 
                                         X1944 
     "республиканцы набрали менее 60% голосов" 
                                         X1948 
     "республиканцы набрали менее 60% голосов" 
                                         X1952 
     "республиканцы набрали менее 60% голосов" 
                                         X1956 
     "республиканцы набрали менее 60% голосов" 
                                         X1960 
     "республиканцы набрали менее 60% голосов" 
                                         X1964 
     "республиканцы набрали менее 60% голосов" 
                                         X1968 
     "республиканцы набрали менее 60% голосов" 
                                         X1972 
"республиканцы набрали высокий поцент голосов" 
                                         X1976 
     "республиканцы набрали менее 60% голосов"

In [30]:
class(ifelse(colMeans(votes.repub,na.rm = TRUE)>60, 
       "республиканцы набрали высокий поцент голосов","–еспубликанцы набрали менее 60% голосов"))

[1] "character"

In [9]:
head(colMeans(votes.repub,na.rm = TRUE))

X1856    X1860    X1864    X1868    X1872    X1876 
39.46850 44.58826 57.88080 54.13697 57.14622 48.49378

Поставим перед собой задачу написать функцию, которая бы считала, на какую сумму продано мячей за определенный день здесь же познакомимся с пакетом lubridate, предназначенного для работы с датами, воспользуемся набором "dat", содержащего данные о продажах в первые дни 2018 года.

__НЕ ЗАБУДЬТЕ УСТАНОВИТЬ РАБОЧУЮ ДИРЕКТОРИЮ, ГДЕ ЛЕЖИТ НУЖНЫЙ ФАЙЛ SESSION->SET WORKING DIRECTORY->CHOOSE DIRECTORY__

In [14]:
getwd()

[1] "C:/YandexDisk/Jupyter/R"

In [15]:
dat<-read.csv("dat.csv")
head(dat)

X d.date     ball price
1 1 2018-01-01 1    90   
2 2 2018-01-01 1    90   
3 3 2018-01-01 1    90   
4 4 2018-01-01 1    90   
5 5 2018-01-01 1    90   
6 6 2018-01-01 1    90

In [16]:
str(dat)

'data.frame':	81 obs. of  4 variables:
 $ X     : int  1 2 3 4 5 6 7 8 9 10 ...
 $ d.date: Factor w/ 3 levels "2018-01-01","2018-01-02",..: 1 1 1 1 1 1 1 1 1 1 ...
 $ ball  : int  1 1 1 1 1 1 1 5 5 5 ...
 $ price : int  90 90 90 90 90 90 90 390 390 390 ...


X - номер покупки

d.date - год-месяц-день

ball - теннисные мячи продавались по одному, а также упакованными по 3 и 5 мячей

price - цены

In [17]:
levels(dat$d.date)

[1] "2018-01-01" "2018-01-02" "2018-01-03"

In [18]:
unique(dat$ball)

[1] 1 5 3

In [19]:
unique(dat$price)

[1]  90 390 200

In [20]:
dim(dat)

[1] 81  4

In [21]:
datn<-dat[, -1]
head(datn, 10)

d.date     ball price
1  2018-01-01 1     90  
2  2018-01-01 1     90  
3  2018-01-01 1     90  
4  2018-01-01 1     90  
5  2018-01-01 1     90  
6  2018-01-01 1     90  
7  2018-01-01 1     90  
8  2018-01-01 5    390  
9  2018-01-01 5    390  
10 2018-01-01 5    390

In [22]:
str(datn)

'data.frame':	81 obs. of  3 variables:
 $ d.date: Factor w/ 3 levels "2018-01-01","2018-01-02",..: 1 1 1 1 1 1 1 1 1 1 ...
 $ ball  : int  1 1 1 1 1 1 1 5 5 5 ...
 $ price : int  90 90 90 90 90 90 90 390 390 390 ...


In [23]:
unique(datn$ball)

[1] 1 5 3

In [24]:
unique(dat$price)

[1]  90 390 200

In [45]:
install.packages("lubridate")

package 'lubridate' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\viv232\AppData\Local\Temp\RtmpCEvopR\downloaded_packages


In [11]:
library(lubridate)

Warning message:
"package 'lubridate' was built under R version 3.6.3"
Attaching package: 'lubridate'

The following objects are masked from 'package:base':

    date, intersect, setdiff, union



In [12]:
?lubridate

lubridate-package          package:lubridate           R Documentation

_D_a_t_e_s _a_n_d _t_i_m_e_s _m_a_d_e _e_a_s_y _w_i_t_h _l_u_b_r_i_d_a_t_e

_D_e_s_c_r_i_p_t_i_o_n:

     Lubridate provides tools that make it easier to parse and
     manipulate dates. These tools are grouped below by common purpose.
     More information about each function can be found in its help
     documentation.

_P_a_r_s_i_n_g _d_a_t_e_s:

     Lubridate's parsing functions read strings into R as POSIXct
     date-time objects. Users should choose the function whose name
     models the order in which the year ('y'), month ('m') and day
     ('d') elements appear the string to be parsed: 'dmy()', 'myd()',
     'ymd()', 'ydm()', 'dym()', 'mdy()', 'ymd_hms()'). A very flexible
     and user friendly parser is provided by 'parse_date_time()'.

     Lubridate can also parse partial dates from strings into Period
     objects with the functions 'hm()', 'hms()' and 'ms()'.

     Lubridate has an inbuilt very fast POSIX parser. Most of the
     'strptime()' formats and various extensions are supported for
     English locales. See 'parse_date_time()' for more details.

_M_a_n_i_p_u_l_a_t_i_n_g _d_a_t_e_s:

     Lubridate distinguishes between moments in time (known as
     'instants()') and spans of time (known as time spans, see
     Timespan). Time spans are further separated into Duration, Period
     and Interval objects.

_I_n_s_t_a_n_t_s:

     Instants are specific moments of time. Date, POSIXct, and POSIXlt
     are the three object classes Base R recognizes as instants.
     'is.Date()' tests whether an object inherits from the Date class.
     'is.POSIXt()' tests whether an object inherits from the POSIXlt or
     POSIXct classes. 'is.instant()' tests whether an object inherits
     from any of the three classes.

     'now()' returns the current system time as a POSIXct object.
     'today()' returns the current system date. For convenience,
     1970-01-01 00:00:00 is saved to origin. This is the instant from
     which POSIXct times are calculated. Try 'unclass(now())' to see
     the numeric structure that underlies POSIXct objects. Each POSIXct
     object is saved as the number of seconds it occurred after
     1970-01-01 00:00:00.

     Conceptually, instants are a combination of measurements on
     different units (i.e, years, months, days, etc.). The individual
     values for these units can be extracted from an instant and set
     with the accessor functions 'second()', 'minute()', 'hour()',
     'day()', 'yday()', 'mday()', 'wday()', 'week()', 'month()',
     'year()', 'tz()', and 'dst()'. Note: the accessor functions are
     named after the singular form of an element. They shouldn't be
     confused with the period helper functions that have the plural
     form of the units as a name (e.g, 'seconds()').

_R_o_u_n_d_i_n_g _d_a_t_e_s:

     Instants can be rounded to a convenient unit using the functions
     'ceiling_date()', 'floor_date()' and 'round_date()'.

_T_i_m_e _z_o_n_e_s:

     Lubridate provides two helper functions for working with time
     zones. 'with_tz()' changes the time zone in which an instant is
     displayed. The clock time displayed for the instant changes, but
     the moment of time described remains the same. 'force_tz()'
     changes only the time zone element of an instant. The clock time
     displayed remains the same, but the resulting instant describes a
     new moment of time.

_T_i_m_e_s_p_a_n_s:

     A timespan is a length of time that may or may not be connected to
     a particular instant. For example, three months is a timespan. So
     is an hour and a half. Base R uses difftime class objects to
     record timespans. However, people are not always consistent in how
     they expect time to behave. Sometimes the passage of time is a
     monotone progression of instants tha

In [25]:
class(datn$d.date)

[1] "factor"

In [26]:
head(datn$d.date)

[1] 2018-01-01 2018-01-01 2018-01-01 2018-01-01 2018-01-01 2018-01-01
Levels: 2018-01-01 2018-01-02 2018-01-03

In [27]:
dayn<-ymd(datn$d.date)
head(dayn)

Warning message in as.POSIXlt.POSIXct(x, tz):
"unable to identify current timezone 'AB@0E0=A:>5 ':
please set environment variable 'TZ'"

[1] "2018-01-01" "2018-01-01" "2018-01-01" "2018-01-01" "2018-01-01"
[6] "2018-01-01"

In [28]:
class(dayn)

[1] "Date"

In [29]:
head(year(dayn))

[1] 2018 2018 2018 2018 2018 2018

In [30]:
head(month(dayn))

[1] 1 1 1 1 1 1

In [31]:
head(day(dayn))

[1] 1 1 1 1 1 1

In [32]:
dat$price

[1]  90  90  90  90  90  90  90 390 390 390 390 390 390 390 390 390 390  90 200
[20] 200  90  90 200  90 200 200  90  90  90  90  90  90  90  90  90  90  90  90
[39]  90  90  90  90  90 390 390 390 390 390 390 390 390 200  90  90  90 200  90
[58]  90 200  90  90  90  90  90  90  90  90  90  90  90 390 390 390 390 390 390
[77] 390 200 200  90 200

In [33]:
datn$price[day(dayn)==1]

[1]  90  90  90  90  90  90  90 390 390 390 390 390 390 390 390 390 390  90 200
[20] 200  90  90 200  90 200 200  90  90  90  90

In [34]:
sum(datn$price[day(dayn)==1])

[1] 6250

### Функции

In [35]:
f_1<-function(d){
  sum(datn$price[day(dayn)==d])
}

In [36]:
f_1(3)

[1] 4410

In [37]:
f_1<-function(d){
  t_1<-sum(datn$price[day(dayn)==d]) #здесь необходимо добавить return(t_1)
  return(t_1)  
}

In [38]:
f_1(3)

[1] 4410

In [39]:
sum(f_1(1),f_1(2),f_1(3))

[1] 16000

In [40]:
sum(datn$price)

[1] 16000

Напишем функцию, которая бы суммировала продажи, но тут будут еще указан месяц

In [58]:
f.2<-function(m,d){sum(datn$price[month(datn$d.date)==m & day(datn$d.date)==d])}

In [59]:
f.2(1,1)

Warning message:
"tz(): Don't know how to compute timezone for object of class factor; returning "UTC". This warning will become an error in the next major version of lubridate."Warning message:
"tz(): Don't know how to compute timezone for object of class factor; returning "UTC". This warning will become an error in the next major version of lubridate."

[1] 6250

In [60]:
f.2(1,3)

Warning message:
"tz(): Don't know how to compute timezone for object of class factor; returning "UTC". This warning will become an error in the next major version of lubridate."Warning message:
"tz(): Don't know how to compute timezone for object of class factor; returning "UTC". This warning will become an error in the next major version of lubridate."

[1] 4410

In [61]:
# ?tz()

In [62]:
# ?force_tz()

In [63]:
f.2(2,1)    #второго месяца в наборе данных нет

Warning message:
"tz(): Don't know how to compute timezone for object of class factor; returning "UTC". This warning will become an error in the next major version of lubridate."Warning message:
"tz(): Don't know how to compute timezone for object of class factor; returning "UTC". This warning will become an error in the next major version of lubridate."

[1] 0

### Циклы for

Посчитаем прибыль магазина по дням. Для этого понадобится цикл for

In [67]:
f_1(1)

[1] 6250

In [68]:
f_1(2)

[1] 5340

In [69]:
for(i in 1:3){
    print(f_1(i))
}

[1] 6250
[1] 5340
[1] 4410


В языке R так делать не надо (сильно снижает производительность). Надо применять операторы группы apply

### Функции apply(), lapply(), sapply(), tapply()

Рассмотрим функции lapply() , sapply()

In [70]:
head(votes.repub)

X1856 X1860 X1864 X1868 X1872 X1876 X1880 X1884 X1888 X1892 ...
Alabama       NA    NA    NA 51.44 53.19 40.02 36.98 38.44 32.28  3.95 ...
Alaska        NA    NA    NA    NA    NA    NA    NA    NA    NA    NA ...
Arizona       NA    NA    NA    NA    NA    NA    NA    NA    NA    NA ...
Arkansas      NA    NA    NA 53.73 52.17 39.88 39.55 40.50 38.07 32.01 ...
California 18.77 32.96 58.63 50.24 56.38 50.88 48.92 52.08 49.95 43.76 ...
Colorado      NA    NA    NA    NA    NA    NA 51.28 54.39 55.31 41.13 ...
           X1940 X1944 X1948 X1952 X1956 X1960 X1964 X1968 X1972 X1976
Alabama    14.34 18.20 19.04 35.02 39.39 41.75 69.5  14.0  72.4  43.48
Alaska        NA    NA    NA    NA    NA 50.94 34.1  45.3  58.1  62.91
Arizona    36.01 40.90 43.82 58.35 60.99 55.52 50.4  54.8  64.7  58.62
Arkansas   20.87 29.84 21.02 43.76 45.82 43.06 43.9  30.8  68.9  34.97
California 41.35 42.99 47.14 56.39 55.40 50.10 40.9  47.8  55.0  50.89
Colorado   50.92 53.21 46.52 60.27 59.49 54.63 38.7  50.5  62.6  55.89

In [71]:
?apply

apply                   package:base                   R Documentation

_A_p_p_l_y _F_u_n_c_t_i_o_n_s _O_v_e_r _A_r_r_a_y _M_a_r_g_i_n_s

_D_e_s_c_r_i_p_t_i_o_n:

     Returns a vector or array or list of values obtained by applying a
     function to margins of an array or matrix.

_U_s_a_g_e:

     apply(X, MARGIN, FUN, ...)
     
_A_r_g_u_m_e_n_t_s:

       X: an array, including a matrix.

  MARGIN: a vector giving the subscripts which the function will be
          applied over.  E.g., for a matrix '1' indicates rows, '2'
          indicates columns, 'c(1, 2)' indicates rows and columns.
          Where 'X' has named dimnames, it can be a character vector
          selecting dimension names.

     FUN: the function to be applied: see 'Details'.  In the case of
          functions like '+', '%*%', etc., the function name must be
          backquoted or quoted.

     ...: optional arguments to 'FUN'.

_D_e_t_a_i_l_s:

     If 'X' is not an array but an object of a class with a non-null
     'dim' value (such as a data frame), 'apply' attempts to coerce it
     to an array via 'as.matrix' if it is two-dimensional (e.g., a data
     frame) or via 'as.array'.

     'FUN' is found by a call to 'match.fun' and typically is either a
     function or a symbol (e.g., a backquoted name) or a character
     string specifying a function to be searched for from the
     environment of the call to 'apply'.

     Arguments in '...' cannot have the same name as any of the other
     arguments, and care may be needed to avoid partial matching to
     'MARGIN' or 'FUN'.  In general-purpose code it is good practice to
     name the first three arguments if '...' is passed through: this
     both avoids partial matching to 'MARGIN' or 'FUN' and ensures that
     a sensible error message is given if arguments named 'X', 'MARGIN'
     or 'FUN' are passed through '...'.

_V_a_l_u_e:

     If each call to 'FUN' returns a vector of length 'n', then 'apply'
     returns an array of dimension 'c(n, dim(X)[MARGIN])' if 'n > 1'.
     If 'n' equals '1', 'apply' returns a vector if 'MARGIN' has length
     1 and an array of dimension 'dim(X)[MARGIN]' otherwise.  If 'n' is
     '0', the result has length 0 but not necessarily the 'correct'
     dimension.

     If the calls to 'FUN' return vectors of different lengths, 'apply'
     returns a list of length 'prod(dim(X)[MARGIN])' with 'dim' set to
     'MARGIN' if this has length greater than one.

     In all cases the result is coerced by 'as.vector' to one of the
     basic vector types before the dimensions are set, so that (for
     example) factor results will be coerced to a character array.

_R_e_f_e_r_e_n_c_e_s:

     Becker, R. A., Chambers, J. M. and Wilks, A. R. (1988) _The New S
     Language_.  Wadsworth & Brooks/Cole.

_S_e_e _A_l_s_o:

     'lapply' and there, 'simplify2array'; 'tapply', and convenience
     functions 'sweep' and 'aggregate'.

_E_x_a_m_p_l_e_s:

     ## Compute row and column sums for a matrix:
     x <- cbind(x1 = 3, x2 = c(4:1, 2:5))
     dimnames(x)[[1]] <- letters[1:8]
     apply(x, 2, mean, trim = .2)
     col.sums <- apply(x, 2, sum)
     row.sums <- apply(x, 1, sum)
     rbind(cbind(x, Rtot = row.sums), Ctot = c(col.sums, sum(col.sums)))
     
     stopifnot( apply(x, 2, is.vector))
     
     ## Sort the columns of a matrix
     apply(x, 2, sort)
     
     ## keeping named dimnames
     names(dimnames(x)) <- c("row", "col")
     x3 <- array(x, dim = c(dim(x),3),
                 dimnames = c(dimnames(x), list(C = paste0("cop.",1:3))))
     identical(x,  apply( x,  2,  identity))
     identical(x3, apply(x3, 2:3, identity))
     
     ##- function with extra args:
     cave <- function(x, c1, c2) c(mean(x[c1]), mean(x[c2]))
     apply(x, 1, cave,  c1 = "x1", c2 = c("x1","x2"))
     
     ma <- matrix(c(1:4, 1, 6:8), nrow = 2)
     ma
     apply(ma, 1, table)  #--> a list of 

In [72]:
class(votes.repub)

[1] "data.frame"

In [74]:
head(apply(votes.repub, 1, mean), 8)   # (apply(votes.repub, 1 по строкам

Alabama      Alaska     Arizona    Arkansas  California    Colorado 
         NA          NA          NA          NA    47.59968          NA 
Connecticut    Delaware 
   51.31452    46.73290

In [83]:
head(apply(votes.repub, 2, mean), 8)   # (apply(votes.repub, 2 по столбцам

X1856 X1860 X1864 X1868 X1872 X1876 X1880 X1884 
   NA    NA    NA    NA    NA    NA    NA    NA

In [75]:
?lapply

lapply                  package:base                   R Documentation

_A_p_p_l_y _a _F_u_n_c_t_i_o_n _o_v_e_r _a _L_i_s_t _o_r _V_e_c_t_o_r

_D_e_s_c_r_i_p_t_i_o_n:

     'lapply' returns a list of the same length as 'X', each element of
     which is the result of applying 'FUN' to the corresponding element
     of 'X'.

     'sapply' is a user-friendly version and wrapper of 'lapply' by
     default returning a vector, matrix or, if 'simplify = "array"', an
     array if appropriate, by applying 'simplify2array()'.  'sapply(x,
     f, simplify = FALSE, USE.NAMES = FALSE)' is the same as 'lapply(x,
     f)'.

     'vapply' is similar to 'sapply', but has a pre-specified type of
     return value, so it can be safer (and sometimes faster) to use.

     'replicate' is a wrapper for the common use of 'sapply' for
     repeated evaluation of an expression (which will usually involve
     random number generation).

     'simplify2array()' is the utility called from 'sapply()' when
     'simplify' is not false and is similarly called from 'mapply()'.

_U_s_a_g_e:

     lapply(X, FUN, ...)
     
     sapply(X, FUN, ..., simplify = TRUE, USE.NAMES = TRUE)
     
     vapply(X, FUN, FUN.VALUE, ..., USE.NAMES = TRUE)
     
     replicate(n, expr, simplify = "array")
     
     simplify2array(x, higher = TRUE)
     
_A_r_g_u_m_e_n_t_s:

       X: a vector (atomic or list) or an 'expression' object.  Other
          objects (including classed objects) will be coerced by
          'base::as.list'.

     FUN: the function to be applied to each element of 'X': see
          'Details'.  In the case of functions like '+', '%*%', the
          function name must be backquoted or quoted.

     ...: optional arguments to 'FUN'.

simplify: logical or character string; should the result be simplified
          to a vector, matrix or higher dimensional array if possible?
          For 'sapply' it must be named and not abbreviated.  The
          default value, 'TRUE', returns a vector or matrix if
          appropriate, whereas if 'simplify = "array"' the result may
          be an 'array' of "rank" (='length(dim(.))') one higher than
          the result of 'FUN(X[[i]])'.

USE.NAMES: logical; if 'TRUE' and if 'X' is character, use 'X' as
          'names' for the result unless it had names already.  Since
          this argument follows '...' its name cannot be abbreviated.

FUN.VALUE: a (generalized) vector; a template for the return value from
          FUN.  See 'Details'.

       n: integer: the number of replications.

    expr: the expression (a language object, usually a call) to
          evaluate repeatedly.

       x: a list, typically returned from 'lapply()'.

  higher: logical; if true, 'simplify2array()' will produce a ("higher
          rank") array when appropriate, whereas 'higher = FALSE' would
          return a matrix (or vector) only.  These two cases correspond
          to 'sapply(*, simplify = "array")' or 'simplify = TRUE',
          respectively.

_D_e_t_a_i_l_s:

     'FUN' is found by a call to 'match.fun' and typically is specified
     as a function or a symbol (e.g., a backquoted name) or a character
     string specifying a function to be searched for from the
     environment of the call to 'lapply'.

     Function 'FUN' must be able to accept as input any of the elements
     of 'X'.  If the latter is an atomic vector, 'FUN' will always be
     passed a length-one vector of the same type as 'X'.

     Arguments in '...' cannot have the same name as any of the other
     arguments, and care may be needed to avoid partial matching to
     'FUN'.  In general-purpose code it is good practice to name the
     first two arguments 'X' and 'FUN' if '...' is passed through: this
     both avoids partial matching to 'FUN' and ensures that a sensible
     error message is given if arguments named 'X' or 'FUN' are passed
     through '...'.

     Simplification 

In [78]:
tail(lapply(votes.repub,sum), 10)

$X1940
[1] NA

$X1944
[1] NA

$X1948
[1] NA

$X1952
[1] NA

$X1956
[1] NA

$X1960
[1] 2497.23

$X1964
[1] 2056.8

$X1968
[1] 2206.3

$X1972
[1] 3135.3

$X1976
[1] 2505.17

In [84]:
class(lapply(votes.repub,sum))

[1] "list"

In [81]:
class(unlist(lapply(votes.repub,sum)))

[1] "numeric"

In [86]:
as.numeric(unlist(lapply(votes.repub,sum)))    # получился sapply

[1]      NA      NA      NA      NA      NA      NA      NA      NA      NA
[10]      NA      NA      NA      NA      NA      NA      NA      NA      NA
[19]      NA      NA      NA      NA      NA      NA      NA      NA 2497.23
[28] 2056.80 2206.30 3135.30 2505.17

In [80]:
tail(sapply(votes.repub,sum), 10)

X1940   X1944   X1948   X1952   X1956   X1960   X1964   X1968   X1972   X1976 
     NA      NA      NA      NA      NA 2497.23 2056.80 2206.30 3135.30 2505.17

In [82]:
class(sapply(votes.repub,sum))

[1] "numeric"

Функция tapply() пименяет нужную функцию к группам. Как она pаботает будет показано на примере датасэта "Orange"

In [91]:
?Orange

Orange                package:datasets                 R Documentation

_G_r_o_w_t_h _o_f _O_r_a_n_g_e _T_r_e_e_s

_D_e_s_c_r_i_p_t_i_o_n:

     The 'Orange' data frame has 35 rows and 3 columns of records of
     the growth of orange trees.

_U_s_a_g_e:

     Orange
     
_F_o_r_m_a_t:

     An object of class 'c("nfnGroupedData", "nfGroupedData",
     "groupedData", "data.frame")' containing the following columns:

     Tree an ordered factor indicating the tree on which the
          measurement is made.  The ordering is according to increasing
          maximum diameter.

     age a numeric vector giving the age of the tree (days since
          1968/12/31)

     circumference a numeric vector of trunk circumferences (mm).  This
          is probably "circumference at breast height", a standard
          measurement in forestry.

_D_e_t_a_i_l_s:

     This dataset was originally part of package 'nlme', and that has
     methods (including for '[', 'as.data.frame', 'plot' and 'print')
     for its grouped-data classes.

_S_o_u_r_c_e:

     Draper, N. R. and Smith, H. (1998), _Applied Regression Analysis
     (3rd ed)_, Wiley (exercise 24.N).

     Pinheiro, J. C. and Bates, D. M. (2000) _Mixed-effects Models in S
     and S-PLUS_, Springer.

_E_x_a_m_p_l_e_s:

     require(stats); require(graphics)
     coplot(circumference ~ age | Tree, data = Orange, show.given = FALSE)
     fm1 <- nls(circumference ~ SSlogis(age, Asym, xmid, scal),
                data = Orange, subset = Tree == 3)
     plot(circumference ~ age, data = Orange, subset = Tree == 3,
          xlab = "Tree age (days since 1968/12/31)",
          ylab = "Tree circumference (mm)", las = 1,
          main = "Orange tree data and fitted model (Tree 3 only)")
     age <- seq(0, 1600, length.out = 101)
     lines(age, predict(fm1, list(age = age)))

In [90]:
str(Orange)

Classes 'nfnGroupedData', 'nfGroupedData', 'groupedData' and 'data.frame':	35 obs. of  3 variables:
 $ Tree         : Ord.factor w/ 5 levels "3"<"1"<"5"<"2"<..: 2 2 2 2 2 2 2 4 4 4 ...
 $ age          : num  118 484 664 1004 1231 ...
 $ circumference: num  30 58 87 115 120 142 145 33 69 111 ...
 - attr(*, "formula")=Class 'formula'  language circumference ~ age | Tree
  .. ..- attr(*, ".Environment")=<environment: R_EmptyEnv> 
 - attr(*, "labels")=List of 2
  ..$ x: chr "Time since December 31, 1968"
  ..$ y: chr "Trunk circumference"
 - attr(*, "units")=List of 2
  ..$ x: chr "(days)"
  ..$ y: chr "(mm)"


In [92]:
head(Orange,12) # окружность измерялась в одинаковом возрасте

Tree age  circumference
1  1     118  30          
2  1     484  58          
3  1     664  87          
4  1    1004 115          
5  1    1231 120          
6  1    1372 142          
7  1    1582 145          
8  2     118  33          
9  2     484  69          
10 2     664 111          
11 2    1004 156          
12 2    1231 172

Набор содержит три переменные:
- номер дерева;
- возраст дерева с 31.12.1968, в днях;
- длина окружности ствола.

Напишем функцию, которая будет считать на сколько изменилась окужность дерева с первой даты замера (118-ый день) до последней даты (1582 день) прежде, введем функции range(), diff()

In [93]:
a<-c(1,2,3,5,8)

In [94]:
diff(a)

[1] 1 1 2 3

In [95]:
range(a)

[1] 1 8

In [96]:
min(a)

[1] 1

In [97]:
max(a)

[1] 8

In [98]:
diff(range(a))

[1] 7

In [99]:
l<-function(x){diff(range(x))}

In [100]:
l(a)

[1] 7

In [103]:
Orange$Tree

[1] 1 1 1 1 1 1 1 2 2 2 2 2 2 2 3 3 3 3 3 3 3 4 4 4 4 4 4 4 5 5 5 5 5 5 5
Levels: 3 < 1 < 5 < 2 < 4

In [102]:
tapply(Orange$circumference,Orange$Tree,l) # 1,2 аргументы функции должны быть одинаковой длины,
                                           #1-й -это то,что разбиваем по подгруппа,2-й- фактор

3   1   5   2   4 
110 115 147 170 182

### Выбор функции с указанием имени пакета (если функции с одинаковыми именами в нескольких пакетах)

In [105]:
search()

[1] ".GlobalEnv"        "package:lubridate" "package:cluster"  
 [4] "jupyter:irkernel"  "package:stats"     "package:graphics" 
 [7] "package:grDevices" "package:utils"     "package:datasets" 
[10] "package:methods"   "Autoloads"         "package:base"

In [115]:
library(effsize)

In [116]:
search()

[1] ".GlobalEnv"        "package:effsize"   "package:lubridate"
 [4] "package:cluster"   "jupyter:irkernel"  "package:stats"    
 [7] "package:graphics"  "package:grDevices" "package:utils"    
[10] "package:datasets"  "package:methods"   "Autoloads"        
[13] "package:base"

In [117]:
?effsize::cohen.d()

cohen.d                package:effsize                 R Documentation

_C_o_h_e_n'_s _d _a_n_d _H_e_d_g_e_s _g _e_f_f_e_c_t _s_i_z_e

_D_e_s_c_r_i_p_t_i_o_n:

     Computes the Cohen's d and Hedges'g effect size statistics.

_U_s_a_g_e:

     cohen.d(d, ...)
     
     ## S3 method for class 'formula'
     cohen.d(formula,data=list(),...)
     
     ## Default S3 method:
     cohen.d(d,f,pooled=TRUE,paired=FALSE,
                        na.rm=FALSE, mu=0, hedges.correction=FALSE,
                        conf.level=0.95,noncentral=FALSE, 
                        within=TRUE, subject=NA, ...)
     
_A_r_g_u_m_e_n_t_s:

       d: a numeric vector giving either the data values (if 'f' is a
          factor) or the treatment group values (if 'f' is a numeric
          vector)

       f: either a factor with two levels or a numeric vector of
          values, if 'NA' a single sample effect size is computed

 formula: a formula of the form 'y ~ f', where 'y' is a numeric
          variable giving the values and 'f' a factor with two levels
          giving the corresponding groups.

          If using a paired computation ('paired=TRUE') it is possible
          to specify the ids of the subjects using the form 'y ~ f |
          Subject(id)' which allow the correct pairing of the pre and
          post values.

          A single sample effect size can be specified with the form 'y
          ~ .'.

    data: an optional matrix or data frame containing the variables in
          the formula 'formula'. By default the variables are taken
          from 'environment(formula)'.

  pooled: a logical indicating whether compute pooled standard
          deviation or the whole sample standard deviation. If
          'pooled=TRUE' (default) pooled sd is used, if 'pooled=FALSE'
          the standard deviation of the the control group (the second
          argument or the one corresponding the the second level of the
          factor) is used instead.

hedges.correction: logical indicating whether apply the Hedges
          correction

conf.level: confidence level of the confidence interval

noncentral: logical indicating whether to use non-central t
          distributions for computing the confidence interval.

  paired: a logical indicating whether to consider the values as
          paired, a warning is issued if 'paired==TRUE' with the
          formula interface and not '| Subject(id)' or with data and
          factor and no 'subject' is provided

  within: indicates whether to compute the effect size using the within
          subject variation, taking into consideration the correlation
          between pre and post samples.

 subject: an array indicating the id of the subject for a paired
          computation, when the formula interface is used it can be
          indicated in the formula by adding '| Subject(id)', where
          'id' is the column in the data that contains and id of the
          subjects to be paired.

      mu: numeric indicating the reference mean for single sample
          effect size.

   na.rm: logical indicating whether 'NA's should be removed before
          computation; if 'paired==TRUE' then all incomplete pairs are
          removed.

     ...: further arguments to be passed to or from methods.

_D_e_t_a_i_l_s:

     When 'f' in the default version is a factor or a character, it
     must have two values and it identifies the two groups to be
     compared. Otherwise (e.g. 'f' is numeric), it is considered as a
     sample to be compare to 'd'.

     In the formula version, 'f' is expected to be a factor, if that is
     not the case it is coherced to a factor and a warning is issued.

     The function computes the value of Cohen's d statistics (Cohen
     1988). If required ('hedges.correction==TRUE') the Hedges g
     statistics is computed instead (Hedges and Holkin, 1985).

     When 'paired' is set, the effect size is computed using the

### Метод Монте-Карло

In [118]:
box_1<-c(rep(1, 5), rep(0, 3))
box_1

[1] 1 1 1 1 1 0 0 0

In [119]:
box_2<-c(rep(1, 5), rep(0, 7))
box_2

[1] 1 1 1 1 1 0 0 0 0 0 0 0

In [121]:
sam.a<-sample(box_1, 2)
sam.a

[1] 0 1

In [122]:
sam.b<-sample(box_2, 4)
sam.b

[1] 0 1 0 1

In [123]:
c(sam.a, sam.b)

[1] 0 1 0 1 0 1

In [125]:
s<-sum(c(sam.a, sam.b))
s

[1] 3

Теперь мы построим этот вектор очень много раз, посчитаем сумму и затем вычислим долю тех значений, которые равны трем. Эта пропорция и будет вероятность того, что мы вытащим ровно три белых мяча.

In [126]:
mc<-function(m, n){
    sam.a<-sample(box_1, m)
    sam.b<-sample(box_2, n)
    s<-sum(c(sam.a, sam.b))
}

In [127]:
m<-replicate(100, mc(2, 4))
m

[1] 3 2 2 2 2 2 4 4 2 2 3 5 3 2 2 4 3 2 3 3 4 2 1 3 4 4 4 3 4 3 2 2 2 3 3 3 2
 [38] 2 3 1 2 3 4 3 4 4 4 4 4 2 5 2 1 1 2 4 3 3 1 2 3 3 4 2 3 4 1 1 3 3 3 4 3 2
 [75] 3 4 4 4 4 4 2 3 3 3 1 3 1 4 2 3 4 4 2 4 2 4 3 1 3 2

In [128]:
mean(m==3)

[1] 0.32

In [129]:
m.1<-replicate(10000, mc(2, 4))
mean(m.1==3)

[1] 0.3669

In [130]:
m.1<-replicate(1000000, mc(2, 4))
mean(m.1==3)

[1] 0.36875

Теоретическое значение: 0.36868

### Пакет dplyr. Оператор %>% (pipe)

In [132]:
library(dplyr)


Attaching package: 'dplyr'

The following objects are masked from 'package:lubridate':

    intersect, setdiff, union

The following objects are masked from 'package:stats':

    filter, lag

The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union



In [133]:
iris %>% dim    # оператор pipe

[1] 150   5

In [135]:
head(iris %>% filter(Species == "versicolor"))

Sepal.Length Sepal.Width Petal.Length Petal.Width Species   
1 7.0          3.2         4.7          1.4         versicolor
2 6.4          3.2         4.5          1.5         versicolor
3 6.9          3.1         4.9          1.5         versicolor
4 5.5          2.3         4.0          1.3         versicolor
5 6.5          2.8         4.6          1.5         versicolor
6 5.7          2.8         4.5          1.3         versicolor

In [138]:
head(iris %>% filter(Species == "versicolor") %>% select(Petal.Length))

Petal.Length
1 4.7         
2 4.5         
3 4.9         
4 4.0         
5 4.6         
6 4.5

In [139]:
head(iris %>% filter(Species == "versicolor") %>% select(Petal.Length) %>% unlist)

Petal.Length1 Petal.Length2 Petal.Length3 Petal.Length4 Petal.Length5 
          4.7           4.5           4.9           4.0           4.6 
Petal.Length6 
          4.5

In [142]:
vc<-as.numeric(iris %>% filter(Species == "versicolor") %>% select(Petal.Length) %>% unlist)
vc

[1] 4.7 4.5 4.9 4.0 4.6 4.5 4.7 3.3 4.6 3.9 3.5 4.2 4.0 4.7 3.6 4.4 4.5 4.1 4.5
[20] 3.9 4.8 4.0 4.9 4.7 4.3 4.4 4.8 5.0 4.5 3.5 3.8 3.7 3.9 5.1 4.5 4.5 4.7 4.4
[39] 4.1 4.0 4.4 4.6 4.0 3.3 4.2 4.2 4.2 4.3 3.0 4.1

Сделаем то же самое, найдем длину лепестков, только для  “virginica”. И с помощью оператора %in% посмотрим, сколько совпадений в длине лепестков двух разных видов цветка.

In [144]:
vg<-as.numeric(iris %>% filter(Species == "virginica") %>% select(Petal.Length) %>% unlist)
vg

[1] 6.0 5.1 5.9 5.6 5.8 6.6 4.5 6.3 5.8 6.1 5.1 5.3 5.5 5.0 5.1 5.3 5.5 6.7 6.9
[20] 5.0 5.7 4.9 6.7 4.9 5.7 6.0 4.8 4.9 5.6 5.8 6.1 6.4 5.6 5.1 5.6 6.1 5.6 5.5
[39] 4.8 5.4 5.6 5.1 5.1 5.9 5.7 5.2 5.0 5.2 5.4 5.1

In [145]:
vc %in% vg

[1] FALSE  TRUE  TRUE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
[13] FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE FALSE  TRUE FALSE  TRUE FALSE
[25] FALSE FALSE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE
[37] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[49] FALSE FALSE

In [146]:
sum(vc %in% vg)

[1] 13

### Сохранение и загрузка данных в формате csv

In [147]:
a<-c(1,2,3)
b<-c(0,0,0)
nm<-paste0(seq(1,3),"_","row")
nm

[1] "1_row" "2_row" "3_row"

In [148]:
df.1<-data.frame(a,b, row.names = nm)
df.1

a b
1_row 1 0
2_row 2 0
3_row 3 0

In [153]:
write.csv(df.1,file = "new_frame.csv")

In [154]:
dir()

[1] "avianHabitat.csv"        "avianHabitat2.csv"      
 [3] "dat.csv"                 "Lesson_1_con.ipynb"     
 [5] "Lesson_2_con.ipynb"      "lesson_2_NEW.R"         
 [7] "Lesson_3_con.ipynb"      "lesson_3_vis_data_1.R"  
 [9] "new_frame.csv"           "november.csv"           
[11] "R.Rproj"                 "R_for_DataScience.ipynb"
[13] "r_poligon.R"             "r_poligon_prv.R"        
[15] "Stepik_R"                "Untitled.ipynb"         
[17] "ДЗ_02.ipynb"

In [155]:
df<-read.csv2("november.csv")
class(df)

[1] "data.frame"

https://www.kaggle.com/sulianova/eda-cardiovascular-data/data

In [171]:
?read.csv

read.table                package:utils                R Documentation

_D_a_t_a _I_n_p_u_t

_D_e_s_c_r_i_p_t_i_o_n:

     Reads a file in table format and creates a data frame from it,
     with cases corresponding to lines and variables to fields in the
     file.

_U_s_a_g_e:

     read.table(file, header = FALSE, sep = "", quote = "\"'",
                dec = ".", numerals = c("allow.loss", "warn.loss", "no.loss"),
                row.names, col.names, as.is = !stringsAsFactors,
                na.strings = "NA", colClasses = NA, nrows = -1,
                skip = 0, check.names = TRUE, fill = !blank.lines.skip,
                strip.white = FALSE, blank.lines.skip = TRUE,
                comment.char = "#",
                allowEscapes = FALSE, flush = FALSE,
                stringsAsFactors = default.stringsAsFactors(),
                fileEncoding = "", encoding = "unknown", text, skipNul = FALSE)
     
     read.csv(file, header = TRUE, sep = ",", quote = "\"",
              dec = ".", fill = TRUE, comment.char = "", ...)
     
     read.csv2(file, header = TRUE, sep = ";", quote = "\"",
               dec = ",", fill = TRUE, comment.char = "", ...)
     
     read.delim(file, header = TRUE, sep = "\t", quote = "\"",
                dec = ".", fill = TRUE, comment.char = "", ...)
     
     read.delim2(file, header = TRUE, sep = "\t", quote = "\"",
                 dec = ",", fill = TRUE, comment.char = "", ...)
     
_A_r_g_u_m_e_n_t_s:

    file: the name of the file which the data are to be read from.
          Each row of the table appears as one line of the file.  If it
          does not contain an _absolute_ path, the file name is
          _relative_ to the current working directory, 'getwd()'.
          Tilde-expansion is performed where supported.  This can be a
          compressed file (see 'file').

          Alternatively, 'file' can be a readable text-mode connection
          (which will be opened for reading if necessary, and if so
          'close'd (and hence destroyed) at the end of the function
          call).  (If 'stdin()' is used, the prompts for lines may be
          somewhat confusing.  Terminate input with a blank line or an
          EOF signal, 'Ctrl-D' on Unix and 'Ctrl-Z' on Windows.  Any
          pushback on 'stdin()' will be cleared before return.)

          'file' can also be a complete URL.  (For the supported URL
          schemes, see the 'URLs' section of the help for 'url'.)

  header: a logical value indicating whether the file contains the
          names of the variables as its first line.  If missing, the
          value is determined from the file format: 'header' is set to
          'TRUE' if and only if the first row contains one fewer field
          than the number of columns.

     sep: the field separator character.  Values on each line of the
          file are separated by this character.  If 'sep = ""' (the
          default for 'read.table') the separator is 'white space',
          that is one or more spaces, tabs, newlines or carriage
          returns.

   quote: the set of quoting characters. To disable quoting altogether,
          use 'quote = ""'.  See 'scan' for the behaviour on quotes
          embedded in quotes.  Quoting is only considered for columns
          read as character, which is all of them unless 'colClasses'
          is specified.

     dec: the character used in the file for decimal points.

numerals: string indicating how to convert numbers whose conversion to
          double precision would lose accuracy, see 'type.convert'.
          Can be abbreviated.  (Applies also to complex-number inputs.)

row.names: a vector of row names.  This can be a vector giving the
          actual row names, or a single number giving the column of the
          table which contains the row names, or character string
          giving the name of the table column containing the row names.

          If there is a header a

In [170]:
read.csv("cardio_train.csv")

Site.Subpoint.VOR.PDB.DBHt.PW.WHt.PE.EHt.PA.AHt.PH.HHt.PL.LHt.PB
1  HortonCreek111;1;3;0;0;0;0;4;3.5;0;0;3;5;0;0;0                  
2  HortonCreek111;2;3;0;0;4;18.5;0;0;0;0;4;5.7;0;0;0               
3  HortonCreek111;3;2;0;0;3;11.3;2;3;0;0;4;4.9;1;0.2;0             
4  HortonCreek111;4;3.5;0;0;4;15.8;0;0;0;0;3;4.2;0;0;0             
5  HortonCreek111;5;0;1;0.3;2;0.5;4;0.4;0;0;2;1.4;5;0.3;0          
6  HortonCreek111;6;0;2;0.7;3;0.5;4;0.5;0;0;1;1.6;3;0.5;1          
7  HortonCreek111;7;0;2;0.3;0;0;5;0.4;0;0;2;1.8;5;0.3;0            
8  HortonCreek112;1;0;1;0.8;0;0;3;0.3;0;0;1;0.9;4;0.4;1            
9  HortonCreek112;2;0;2;0.5;0;0;3;0.4;0;0;2;0.7;3;0.1;1            
10 HortonCreek112;3;0;1;0.1;0;0;3;0.2;0;0;1;0.5;4;0.1;2            
11 HortonCreek112;4;0;2;0.3;0;0;3;0.2;0;0;1;0.5;4;0.1;1            
12 HortonCreek112;5;0;0;0;0;0;4;0.7;0;0;1;0.5;2;0.1;2              
13 HortonCreek112;6;0;0;0;0;0;2;0.3;0;0;2;0.7;3;0.3;3              
14 HortonCreek112;7;0;0;0;2;0.3;3;0.3;0;0;2;0.8;4;0.4;2            
15 HortonCreek112;8;0;3;0.7;0;0;0;0;0;0;1;3.1;5;0.5;0              
16 HortonCreek112;9;0;0;0;0;0;2;0.2;0;0;2;0.7;1;0.3;4

In [158]:
head(read.csv("cardio_train.csv",sep=";"))

Site           Subpoint VOR PDB DBHt PW WHt  PE EHt PA AHt PH HHt PL LHt PB
1 HortonCreek111 1        3.0 0   0.0  0   0.0 4  3.5 0  0   3  5.0 0  0.0 0 
2 HortonCreek111 2        3.0 0   0.0  4  18.5 0  0.0 0  0   4  5.7 0  0.0 0 
3 HortonCreek111 3        2.0 0   0.0  3  11.3 2  3.0 0  0   4  4.9 1  0.2 0 
4 HortonCreek111 4        3.5 0   0.0  4  15.8 0  0.0 0  0   3  4.2 0  0.0 0 
5 HortonCreek111 5        0.0 1   0.3  2   0.5 4  0.4 0  0   2  1.4 5  0.3 0 
6 HortonCreek111 6        0.0 2   0.7  3   0.5 4  0.5 0  0   1  1.6 3  0.5 1

In [168]:
head(read.csv2("cardio_train.csv"))

Site           Subpoint VOR PDB DBHt PW WHt  PE EHt PA AHt PH HHt PL LHt PB
1 HortonCreek111 1        3   0   0    0  0    4  3.5 0  0   3  5   0  0   0 
2 HortonCreek111 2        3   0   0    4  18.5 0  0   0  0   4  5.7 0  0   0 
3 HortonCreek111 3        2   0   0    3  11.3 2  3   0  0   4  4.9 1  0.2 0 
4 HortonCreek111 4        3.5 0   0    4  15.8 0  0   0  0   3  4.2 0  0   0 
5 HortonCreek111 5        0   1   0.3  2  0.5  4  0.4 0  0   2  1.4 5  0.3 0 
6 HortonCreek111 6        0   2   0.7  3  0.5  4  0.5 0  0   1  1.6 3  0.5 1

In [159]:
read.csv("cardio_train.csv", sep=";", nrow=5, row.names = paste0("row", seq(1,5))) # сколько строк хотим прочитать, задать имя строки

Site           Subpoint VOR PDB DBHt PW WHt  PE EHt PA AHt PH HHt PL LHt
row1 HortonCreek111 1        3.0 0   0.0  0   0.0 4  3.5 0  0   3  5.0 0  0.0
row2 HortonCreek111 2        3.0 0   0.0  4  18.5 0  0.0 0  0   4  5.7 0  0.0
row3 HortonCreek111 3        2.0 0   0.0  3  11.3 2  3.0 0  0   4  4.9 1  0.2
row4 HortonCreek111 4        3.5 0   0.0  4  15.8 0  0.0 0  0   3  4.2 0  0.0
row5 HortonCreek111 5        0.0 1   0.3  2   0.5 4  0.4 0  0   2  1.4 5  0.3
     PB
row1 0 
row2 0 
row3 0 
row4 0 
row5 0

In [160]:
read.csv("cardio_train.csv", sep=";", nrow=5, row.names = paste0("row", seq(1,5)), header= FALSE)

V1             V2       V3  V4  V5   V6 V7   V8 V9  V10 V11 V12 V13 V14
row1 Site           Subpoint VOR PDB DBHt PW WHt  PE EHt PA  AHt PH  HHt PL 
row2 HortonCreek111 1        3   0   0    0  0    4  3.5 0   0   3   5   0  
row3 HortonCreek111 2        3   0   0    4  18.5 0  0   0   0   4   5.7 0  
row4 HortonCreek111 3        2   0   0    3  11.3 2  3   0   0   4   4.9 1  
row5 HortonCreek111 4        3.5 0   0    4  15.8 0  0   0   0   3   4.2 0  
     V15 V16
row1 LHt PB 
row2 0   0  
row3 0   0  
row4 0.2 0  
row5 0   0

## Загрузка пакета для чтения данных в фомате excel

In [162]:
library(readxl)

In [163]:
read_xls("book_1.xls")

a 1 2 3
1 b 1 2 3
2 c 1 2 3
3 d 1 2 3
4 e 1 2 3
5 f 1 2 3

In [165]:
install.packages("rio")

also installing the dependencies 'zip', 'openxlsx'



package 'zip' successfully unpacked and MD5 sums checked
package 'openxlsx' successfully unpacked and MD5 sums checked
package 'rio' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\viv232\AppData\Local\Temp\Rtmpmq2kZ4\downloaded_packages


In [166]:
library(rio)

Warning message:
"package 'rio' was built under R version 3.6.3"

In [167]:
?rio

rio                    package:rio                     R Documentation

_A _S_w_i_s_s-_A_r_m_y _K_n_i_f_e _f_o_r _D_a_t_a _I/_O

_D_e_s_c_r_i_p_t_i_o_n:

     The aim of rio is to make data file input and output as easy as
     possible. 'export' and 'import' serve as a Swiss-army knife for
     painless data I/O for data from almost any file format by
     inferring the data structure from the file extension, natively
     reading web-based data sources, setting reasonable defaults for
     import and export, and relying on efficient data import and export
     packages. An additional convenience function, 'convert', provides
     a simple method for converting between file types.

     Note that some of rio's functionality is provided by 'Suggests'
     dependendencies, meaning they are not installed by default. Use
     'install_formats' to make sure these packages are available for
     use.

_R_e_f_e_r_e_n_c_e_s:

     GREA provides an RStudio add-in to import data using rio.

_S_e_e _A_l_s_o:

     'import', 'import_list', 'export', 'convert', 'install_formats'

_E_x_a_m_p_l_e_s:

     # export
     library("datasets")
     export(mtcars, "mtcars.csv") # comma-separated values
     export(mtcars, "mtcars.rds") # R serialized
     export(mtcars, "mtcars.sav") # SPSS
     
     # import
     x <- import("mtcars.csv")
     y <- import("mtcars.rds")
     z <- import("mtcars.sav")
     
     # convert
     convert("mtcars.sav", "mtcars.dta")
     
     # cleanup
     unlink(c("mtcars.csv", "mtcars.rds", "mtcars.sav", "mtcars.dta"))

Cравним базовые функции и функции пакета "рио"

In [173]:
head(import("cardio_train.csv"))

Site           Subpoint VOR PDB DBHt PW WHt  PE EHt PA AHt PH HHt PL LHt PB
1 HortonCreek111 1        3.0 0   0.0  0   0.0 4  3.5 0  0   3  5.0 0  0.0 0 
2 HortonCreek111 2        3.0 0   0.0  4  18.5 0  0.0 0  0   4  5.7 0  0.0 0 
3 HortonCreek111 3        2.0 0   0.0  3  11.3 2  3.0 0  0   4  4.9 1  0.2 0 
4 HortonCreek111 4        3.5 0   0.0  4  15.8 0  0.0 0  0   3  4.2 0  0.0 0 
5 HortonCreek111 5        0.0 1   0.3  2   0.5 4  0.4 0  0   2  1.4 5  0.3 0 
6 HortonCreek111 6        0.0 2   0.7  3   0.5 4  0.5 0  0   1  1.6 3  0.5 1

In [174]:
import("book_1.xls")

a 1 2 3
1 b 1 2 3
2 c 1 2 3
3 d 1 2 3
4 e 1 2 3
5 f 1 2 3

### Загрузка данных из сети

https://biometry.nci.nih.gov/cdas/datasets/nlst/

In [175]:
library(rvest)

Loading required package: xml2
Registered S3 method overwritten by 'rvest':
  method            from
  read_xml.response xml2


In [176]:
url_1<-paste0("https://biometry.nci.nih.gov/cdas/datasets/nlst/")
url_1

[1] "https://biometry.nci.nih.gov/cdas/datasets/nlst/"

In [177]:
htl<-read_html(url_1)    # чтение HTML или XML
htl

{xml_document}
<html lang="en">
[1] <head>\n<meta http-equiv="Content-Type" content="text/html; charset=utf-8 ...
[2] <body class="">\n        <a href="#maincontent" class="sr-only sr-only-fo ...

In [178]:
tabl<-htl %>% html_nodes("table")
tabl

{xml_nodeset (2)}
[1] <table class="table table-striped table-bordered">\n<tr>\n<td class="nowr ...
[2] <table class="table table-striped table-bordered" style="width: auto; min ...

In [179]:
tabl<-tabl[[1]] %>% html_table    # загружаем нужную нам таблицу
head(tabl)

X1                                                         
1 Data Dictionary\n            \n            (PDF - 553.4 KB)
2 Data Dictionary\n            \n            (PDF - 210.0 KB)
3 Data Dictionary\n            \n            (PDF - 189.9 KB)
4 Data Dictionary\n            \n            (PDF - 178.2 KB)
5 Data Dictionary\n            \n            (PDF - 169.0 KB)
6 Data Dictionary\n            \n            (PDF - 175.5 KB)
  X2                                                                                                                                                                                                                                                                                                                                                                                                                                   
1 1.\n                \n                    The Participant dataset is a comprehensive dataset that contains all the NLST study data needed for most analyses of lung cancer screening, incidence, and mortality.  The dataset contains one record for each of the ~53,500 participants in NLST.                                                                                                                                       
2 2.\n                \n                    The Spiral CT Screening dataset (~75,100, one record per CT screen) contains information from the Spiral CT screening exams. This includes technical parameters, reconstruction filter(s), reader ID, and recommendations for diagnostic follow-up.                                                                                                                                        
3 3.\n                \n                    The Chest X-Ray Screening dataset (~73,500, one record per X-Ray screen) contains information from the Chest X-Ray screening exams. This includes technical parameters, reader ID, and recommendations for diagnostic follow-up.                                                                                                                                                           
4 4.\n                \n                    The Spiral CT Abnormalities dataset (~177,500, one record per abnormality on CT) contains information about each abnormality observed on the Spiral CT screening exams.                                                                                                                                                                                                                    
5 5.\n                \n                    The Chest X-Ray Abnormalities dataset (~47,200, one record per abnormality on X-Ray) contains information about each abnormality observed on the Chest X-Ray screening exams.                                                                                                                                                                                                              
6 6.\n                \n                    The Spiral CT Comparison Read Abnormalities dataset (~31,000, one record per abnormality on CT) contains information about two types of abnormalities observed on the comparison read of CT exams: (a) all non-calcified nodules / masses >= 4mm in diameter; (b) other abnormalities deemed significant by the radiologist. Information about change in size and attenuation is available.

In [180]:
class(tabl)

[1] "data.frame"

In [181]:
tabl<-tabl %>% setNames(c("data", "description"))
head(tabl)

data                                                       
1 Data Dictionary\n            \n            (PDF - 553.4 KB)
2 Data Dictionary\n            \n            (PDF - 210.0 KB)
3 Data Dictionary\n            \n            (PDF - 189.9 KB)
4 Data Dictionary\n            \n            (PDF - 178.2 KB)
5 Data Dictionary\n            \n            (PDF - 169.0 KB)
6 Data Dictionary\n            \n            (PDF - 175.5 KB)
  description                                                                                                                                                                                                                                                                                                                                                                                                                          
1 1.\n                \n                    The Participant dataset is a comprehensive dataset that contains all the NLST study data needed for most analyses of lung cancer screening, incidence, and mortality.  The dataset contains one record for each of the ~53,500 participants in NLST.                                                                                                                                       
2 2.\n                \n                    The Spiral CT Screening dataset (~75,100, one record per CT screen) contains information from the Spiral CT screening exams. This includes technical parameters, reconstruction filter(s), reader ID, and recommendations for diagnostic follow-up.                                                                                                                                        
3 3.\n                \n                    The Chest X-Ray Screening dataset (~73,500, one record per X-Ray screen) contains information from the Chest X-Ray screening exams. This includes technical parameters, reader ID, and recommendations for diagnostic follow-up.                                                                                                                                                           
4 4.\n                \n                    The Spiral CT Abnormalities dataset (~177,500, one record per abnormality on CT) contains information about each abnormality observed on the Spiral CT screening exams.                                                                                                                                                                                                                    
5 5.\n                \n                    The Chest X-Ray Abnormalities dataset (~47,200, one record per abnormality on X-Ray) contains information about each abnormality observed on the Chest X-Ray screening exams.                                                                                                                                                                                                              
6 6.\n                \n                    The Spiral CT Comparison Read Abnormalities dataset (~31,000, one record per abnormality on CT) contains information about two types of abnormalities observed on the comparison read of CT exams: (a) all non-calcified nodules / masses >= 4mm in diameter; (b) other abnormalities deemed significant by the radiologist. Information about change in size and attenuation is available.

Рассмотрим отдельный столбец датафрэйма

In [182]:
tabl$data

[1] "Data Dictionary\n            \n            (PDF - 553.4 KB)"
 [2] "Data Dictionary\n            \n            (PDF - 210.0 KB)"
 [3] "Data Dictionary\n            \n            (PDF - 189.9 KB)"
 [4] "Data Dictionary\n            \n            (PDF - 178.2 KB)"
 [5] "Data Dictionary\n            \n            (PDF - 169.0 KB)"
 [6] "Data Dictionary\n            \n            (PDF - 175.5 KB)"
 [7] "Data Dictionary\n            \n            (PDF - 175.3 KB)"
 [8] "Data Dictionary\n            \n            (PDF - 177.6 KB)"
 [9] "Data Dictionary\n            \n            (PDF - 179.3 KB)"
[10] "Data Dictionary\n            \n            (PDF - 222.4 KB)"
[11] "Data Dictionary\n            \n            (PDF - 171.9 KB)"
[12] "Data Dictionary\n            \n            (PDF - 160.6 KB)"
[13] "Data Dictionary\n            \n            (PDF - 165.8 KB)"
[14] "Data Dictionary\n            \n            (PDF - 159.5 KB)"
[15] "Data Dictionary\n            \n            (PDF - 197.5 KB)"

In [183]:
class(tabl$data)

[1] "character"

In [185]:
t1<-sub(pattern = "\n            \n            ", replacement=" ", tabl$data)
t1

[1] "Data Dictionary (PDF - 553.4 KB)" "Data Dictionary (PDF - 210.0 KB)"
 [3] "Data Dictionary (PDF - 189.9 KB)" "Data Dictionary (PDF - 178.2 KB)"
 [5] "Data Dictionary (PDF - 169.0 KB)" "Data Dictionary (PDF - 175.5 KB)"
 [7] "Data Dictionary (PDF - 175.3 KB)" "Data Dictionary (PDF - 177.6 KB)"
 [9] "Data Dictionary (PDF - 179.3 KB)" "Data Dictionary (PDF - 222.4 KB)"
[11] "Data Dictionary (PDF - 171.9 KB)" "Data Dictionary (PDF - 160.6 KB)"
[13] "Data Dictionary (PDF - 165.8 KB)" "Data Dictionary (PDF - 159.5 KB)"
[15] "Data Dictionary (PDF - 197.5 KB)"

Предположим, что нам нужно оставить только "Dictionary"

In [186]:
t1[1]    # для простоты понимания рассмотрим работу функций на первом элементе текстового вектора

[1] "Data Dictionary (PDF - 553.4 KB)"

In [188]:
strsplit(t1[1], split = NULL)

[[1]]
 [1] "D" "a" "t" "a" " " "D" "i" "c" "t" "i" "o" "n" "a" "r" "y" " " "(" "P" "D"
[20] "F" " " "-" " " "5" "5" "3" "." "4" " " "K" "B" ")"

In [189]:
class(strsplit(t1[1], split = NULL))

[1] "list"

In [190]:
length(strsplit(t1[1], split = NULL))    # не покажет количество символов

[1] 1

In [191]:
vec<-unlist(strsplit(t1[1], split = NULL))
vec

[1] "D" "a" "t" "a" " " "D" "i" "c" "t" "i" "o" "n" "a" "r" "y" " " "(" "P" "D"
[20] "F" " " "-" " " "5" "5" "3" "." "4" " " "K" "B" ")"

In [192]:
class(vec)

[1] "character"

In [193]:
length(vec)

[1] 32

In [195]:
vec[1:10]

[1] "D" "a" "t" "a" " " "D" "i" "c" "t" "i"

In [196]:
vec[1:15]

[1] "D" "a" "t" "a" " " "D" "i" "c" "t" "i" "o" "n" "a" "r" "y"

In [197]:
substr(t1[1], start = 6, stop = 15)

[1] "Dictionary"

In [198]:
Dictionary_number<-substr(t1, 6, 15)
Dictionary_number

[1] "Dictionary" "Dictionary" "Dictionary" "Dictionary" "Dictionary"
 [6] "Dictionary" "Dictionary" "Dictionary" "Dictionary" "Dictionary"
[11] "Dictionary" "Dictionary" "Dictionary" "Dictionary" "Dictionary"

In [200]:
Dictionary_final<-paste0(Dictionary_number, set = "_", 1:15)
Dictionary_final

[1] "Dictionary_1"  "Dictionary_2"  "Dictionary_3"  "Dictionary_4" 
 [5] "Dictionary_5"  "Dictionary_6"  "Dictionary_7"  "Dictionary_8" 
 [9] "Dictionary_9"  "Dictionary_10" "Dictionary_11" "Dictionary_12"
[13] "Dictionary_13" "Dictionary_14" "Dictionary_15"

In [201]:
description<-tabl[, 2]
description[1]

[1] "1.\n                \n                    The Participant dataset is a comprehensive dataset that contains all the NLST study data needed for most analyses of lung cancer screening, incidence, and mortality.  The dataset contains one record for each of the ~53,500 participants in NLST."

In [208]:
description_new<-sub(".\n                \n                    ", replacement = "", description)
head(description_new)

[1] "1The Participant dataset is a comprehensive dataset that contains all the NLST study data needed for most analyses of lung cancer screening, incidence, and mortality.  The dataset contains one record for each of the ~53,500 participants in NLST."                                                                                                                                       
[2] "2The Spiral CT Screening dataset (~75,100, one record per CT screen) contains information from the Spiral CT screening exams. This includes technical parameters, reconstruction filter(s), reader ID, and recommendations for diagnostic follow-up."                                                                                                                                        
[3] "3The Chest X-Ray Screening dataset (~73,500, one record per X-Ray screen) contains information from the Chest X-Ray screening exams. This includes technical parameters, reader ID, and recommendations for diagnostic follow-up."                                                                                                                                                           
[4] "4The Spiral CT Abnormalities dataset (~177,500, one record per abnormality on CT) contains information about each abnormality observed on the Spiral CT screening exams."                                                                                                                                                                                                                    
[5] "5The Chest X-Ray Abnormalities dataset (~47,200, one record per abnormality on X-Ray) contains information about each abnormality observed on the Chest X-Ray screening exams."                                                                                                                                                                                                              
[6] "6The Spiral CT Comparison Read Abnormalities dataset (~31,000, one record per abnormality on CT) contains information about two types of abnormalities observed on the comparison read of CT exams: (a) all non-calcified nodules / masses >= 4mm in diameter; (b) other abnormalities deemed significant by the radiologist. Information about change in size and attenuation is available."

In [209]:
description_new<-substr(description_new, start = 3, stop = nchar(description_new))
head(description_new)

[1] "he Participant dataset is a comprehensive dataset that contains all the NLST study data needed for most analyses of lung cancer screening, incidence, and mortality.  The dataset contains one record for each of the ~53,500 participants in NLST."                                                                                                                                       
[2] "he Spiral CT Screening dataset (~75,100, one record per CT screen) contains information from the Spiral CT screening exams. This includes technical parameters, reconstruction filter(s), reader ID, and recommendations for diagnostic follow-up."                                                                                                                                        
[3] "he Chest X-Ray Screening dataset (~73,500, one record per X-Ray screen) contains information from the Chest X-Ray screening exams. This includes technical parameters, reader ID, and recommendations for diagnostic follow-up."                                                                                                                                                           
[4] "he Spiral CT Abnormalities dataset (~177,500, one record per abnormality on CT) contains information about each abnormality observed on the Spiral CT screening exams."                                                                                                                                                                                                                    
[5] "he Chest X-Ray Abnormalities dataset (~47,200, one record per abnormality on X-Ray) contains information about each abnormality observed on the Chest X-Ray screening exams."                                                                                                                                                                                                              
[6] "he Spiral CT Comparison Read Abnormalities dataset (~31,000, one record per abnormality on CT) contains information about two types of abnormalities observed on the comparison read of CT exams: (a) all non-calcified nodules / masses >= 4mm in diameter; (b) other abnormalities deemed significant by the radiologist. Information about change in size and attenuation is available."

In [211]:
description<-paste0('T', description_new[1:9])
New_Description2<-c(description, description_new[10:15])
head(New_Description2)

[1] "The Participant dataset is a comprehensive dataset that contains all the NLST study data needed for most analyses of lung cancer screening, incidence, and mortality.  The dataset contains one record for each of the ~53,500 participants in NLST."                                                                                                                                       
[2] "The Spiral CT Screening dataset (~75,100, one record per CT screen) contains information from the Spiral CT screening exams. This includes technical parameters, reconstruction filter(s), reader ID, and recommendations for diagnostic follow-up."                                                                                                                                        
[3] "The Chest X-Ray Screening dataset (~73,500, one record per X-Ray screen) contains information from the Chest X-Ray screening exams. This includes technical parameters, reader ID, and recommendations for diagnostic follow-up."                                                                                                                                                           
[4] "The Spiral CT Abnormalities dataset (~177,500, one record per abnormality on CT) contains information about each abnormality observed on the Spiral CT screening exams."                                                                                                                                                                                                                    
[5] "The Chest X-Ray Abnormalities dataset (~47,200, one record per abnormality on X-Ray) contains information about each abnormality observed on the Chest X-Ray screening exams."                                                                                                                                                                                                              
[6] "The Spiral CT Comparison Read Abnormalities dataset (~31,000, one record per abnormality on CT) contains information about two types of abnormalities observed on the comparison read of CT exams: (a) all non-calcified nodules / masses >= 4mm in diameter; (b) other abnormalities deemed significant by the radiologist. Information about change in size and attenuation is available."

__Новые обработанные переменные объединим в датафрэйм__

In [212]:
data.frame(tolower(Dictionary_final), New_Description2)    # изменим на строчную букву первый столбец

tolower.Dictionary_final.
1  dictionary_1             
2  dictionary_2             
3  dictionary_3             
4  dictionary_4             
5  dictionary_5             
6  dictionary_6             
7  dictionary_7             
8  dictionary_8             
9  dictionary_9             
10 dictionary_10            
11 dictionary_11            
12 dictionary_12            
13 dictionary_13            
14 dictionary_14            
15 dictionary_15            
   New_Description2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
1  The Participant dataset is a comprehensive dataset that contains all the NLST study data needed for most analyses of lung cancer screening, incidence, and mortality.  The dataset contains one record for each of the ~53,500 participants in NLST.                                                                                                                                                                                                                                                                                     
2  The Spiral CT Screening dataset (~75,100, one record per CT screen) contains information from the Spiral CT screening exams. This includes technical parameters, reconstruction filter(s), reader ID, and recommendations for diagnostic follow-up.                                                                                                                                                                                                                                                                                      
3  The Chest X-Ray Screening dataset (~73,500, one record per X-Ray screen) contains information from the Chest X-Ray screening exams. This includes technical parameters, reader ID, and recommendations for diagnostic follow-up.                                                                                                                                                                                                                                                                                                         
4  The Spiral CT Abnormalities dataset (~177,500, one record per abnormality on CT) contains information about each abnormality observed on the Spiral CT screening exams.                                                                                                                                                                                                                                                                                                                                                                  
5  The Chest X-Ray Abnormalities dataset (~47,200, one record per abnormality on X-Ray) contains information about each abnormality observed on the Chest X-Ray screening exams.                                                                                                                                                                                                                                                                                                                                                            
6  The Spiral CT Comparison Read Abnormalities dataset (~31,000, one record per abnormality on CT) contains information about two types of abnormalities observed on the comparison read of CT exams: (a) all non-calcified nodules / masses >= 4mm in diameter; (b) other abnormalities deemed significant by the radiologist. Information about change in size and attenuation is available.       

__Воспользуемся функцией data_frame из пакета dplyr <br>Эта функция делает данные читаемыми, легкими для восприятия__

In [ ]:
library(dplyr)

In [215]:
data_frame(tolower(Dictionary_final), New_Description2)

Warning message:
"`data_frame()` is deprecated, use `tibble()`.
This warning is displayed once per session."

tolower(Dictionary_final)
1  dictionary_1             
2  dictionary_2             
3  dictionary_3             
4  dictionary_4             
5  dictionary_5             
6  dictionary_6             
7  dictionary_7             
8  dictionary_8             
9  dictionary_9             
10 dictionary_10            
11 dictionary_11            
12 dictionary_12            
13 dictionary_13            
14 dictionary_14            
15 dictionary_15            
   New_Description2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
1  The Participant dataset is a comprehensive dataset that contains all the NLST study data needed for most analyses of lung cancer screening, incidence, and mortality.  The dataset contains one record for each of the ~53,500 participants in NLST.                                                                                                                                                                                                                                                                                     
2  The Spiral CT Screening dataset (~75,100, one record per CT screen) contains information from the Spiral CT screening exams. This includes technical parameters, reconstruction filter(s), reader ID, and recommendations for diagnostic follow-up.                                                                                                                                                                                                                                                                                      
3  The Chest X-Ray Screening dataset (~73,500, one record per X-Ray screen) contains information from the Chest X-Ray screening exams. This includes technical parameters, reader ID, and recommendations for diagnostic follow-up.                                                                                                                                                                                                                                                                                                         
4  The Spiral CT Abnormalities dataset (~177,500, one record per abnormality on CT) contains information about each abnormality observed on the Spiral CT screening exams.                                                                                                                                                                                                                                                                                                                                                                  
5  The Chest X-Ray Abnormalities dataset (~47,200, one record per abnormality on X-Ray) contains information about each abnormality observed on the Chest X-Ray screening exams.                                                                                                                                                                                                                                                                                                                                                            
6  The Spiral CT Comparison Read Abnormalities dataset (~31,000, one record per abnormality on CT) contains information about two types of abnormalities observed on the comparison read of CT exams: (a) all non-calcified nodules / masses >= 4mm in diameter; (b) other abnormalities deemed significant by the radiologist. Information about change in size and attenuation is available.       